### PR TITLE
BHoM_Adapter: allow push multiple Types at the same time by gathering dependency order

### LINKED ISSUE #331: Allow push of multiple types at the same time
#### Description:

As discussed with @IsakNaslundBh, @enricoantolini, and others, an useful workflow that we could support for the BHoM adapters -- in particular structural adapters -- is to allow the push of different types at the same time, dependencies included. This would be particularly helpful for models converted via the [Karamba3D_Toolkit](https://github.com/BHoM/Karamba3D_Toolkit).

Currently the dependencyTypes logic falls within the `FullCRUD` method, which means that dependencies are handled on a per-type basis. Therefore, objects of certain types like `BarLoads` and `Bars` need a 2-step process, where Bars are pushed first, their AdapterId is assigned, and then they can be used to define BarLoads.

This can be improved by moving the dependency handling logic outside of the FullCRUD, before it gets called. We can collect and order the objects prior to them being sent to the FullCRUD based on the DependencyTypes (diagram credit @IsakNaslundBh):
image

Further, in order to allow a faster processing of this workflow, we can add things like:
- caching objects that have been pushed, then matching them with object dependants by hash or other means (adding the exception of AdpterID fragment)
- avoid reading objects whose ID is already known (targeted by https://github.com/BHoM/BHoM_Adapter/issues/329)

### PR DESCRIPTION
**Pull request by alelom**
### Depends on

https://github.com/BHoM/BHoM_Engine/pull/2974

   
### Issues addressed by this PR

Closes #331 

### Test files

Use the full testing procedure for Adapters on this PR. 

Additional tests are temporarily provided in the form of Unit Test within the solution.
I found it extremely useful to develop this functionality by preparing tests before developing, because I could run the whole adapter workflow, debug and live-edit while debugging. I wish we could keep them where they are but, because currently this is non-compliant, they will likely be moved out in a dedicated toolkit, similarly to what we do with DiffingTests_prototypes.

### Additional comments

Because this feature is needed to fully leverage the work done by @enricoantolini in the Karamba_Toolkit, I am putting it as a high priority feature, to be merged as soon as possible. However, in depth testing is needed on this one. 

Additionally, some ancillary features are needed to avoid some slowdown introduced by this PR, like https://github.com/BHoM/BHoM_Adapter/issues/329, which should be mergeable soon after this gets merged.

### CODE REVIEW

**Comment by al-fisher** [comment]
Nice!
Quick comment on the Unit Tests - think good case to be made to keep closer to code for advanced framework projects like this!
The core framework code (which demands necessarily more complex functionality - and I would include in this the Adaptor and UI repos, and perhaps core functionality like versioning and serialisation etc.) should naturally have robust unit tests - and should be stored where most logical to ease development, testing and review.
This is perhaps converse to our “discipline engines” which we should continue to strive to be kept as “simple as possible”, relying heavily on our data set driven unit tests and for which the compliance rules were largely originally derived I feel. 

Worth opening a discussion on refining compliance for the advanced cases like this @alelom to enanle gathering thoughts and consensus ?

File: Adapter_Engine/Query/BaseTypes.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+        public static Dictionary<Type, List<Type>> m_cachedBaseTypes = new Dictionary<Type, List<Type>>();
+
+        [Description("Gets all implemented interfaces and any base type of a given type.")]
+        public static List<Type> BaseTypes(this Type type)
```
Comment (FraserGreenroyd) [inline Adapter_Engine/Query/BaseTypes.cs:44]: I think this should sit in the Reflection_Engine, similar to `CommonBaseType` that we already have. This is code asking questions about itself, so is better placed there than here.

File: Adapter_Engine/Query/BaseTypes.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        {
+            List<Type> baseTypes = new List<Type>();
```
Comment (FraserGreenroyd) [inline Adapter_Engine/Query/BaseTypes.cs:46]: I'd put this after the null check - saves creating a list in memory that isn't needed.

File: Adapter_Engine/Query/BaseTypes.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+            if (type == null)
+                return baseTypes;
+
+            if (m_cachedBaseTypes.TryGetValue(type, out baseTypes))
+                return baseTypes;
+            else
+                baseTypes = new List<Type>();
```
Comment (FraserGreenroyd) [inline Adapter_Engine/Query/BaseTypes.cs:54]: Based on the code above, this `else` isn't doing anything extra so can be removed

File: Adapter_Engine/Adapter_Engine.csproj [hunk 4]
```diff
@@ -16,9 +16,13 @@
   </PropertyGroup>
 
   <Target Name="PostBuild" AfterTargets="PostBuildEvent">
-    <Exec Command="xcopy &quot;$(TargetDir)$(TargetFileName)&quot;  &quot;C:\ProgramData\BHoM\Assemblies&quot; /Y" />
+    <Exec Command="xcopy &quot;$(TargetDir)$(TargetFileName)&quot;  &quot;C:\ProgramData\BHoM\Assemblies&quot; /Y&#xD;&#xA;&#xD;&#xA;Exit 0" />
```
Comment (FraserGreenroyd) [inline Adapter_Engine/Adapter_Engine.csproj:19]: Umm 👀

File: Adapter_Engine/Query/GetDependencyRecursive.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+                    gatheredDependecies[depObj.Key] = depObj.Value.Cast<IBHoMObject>().ToList();
+
+                GetDependencyObjectsRecursive(depObj.Value as dynamic, gatheredDependecies, adapter);
+            }
+        }
+    }
+}
+
+
+
```
Comment (FraserGreenroyd) [inline Adapter_Engine/Query/GetDependencyRecursive.cs:85]:
```suggestion
```

File: Adapter_Engine/Query/GetDependencyRecursive.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                MethodInfo enumCastMethod_specificType = typeof(Enumerable).GetMethod("Cast").MakeGenericMethod(new[] { typeGroup.Key });
+                dynamic objList_specificType = enumCastMethod_specificType.Invoke(typeGroup, new object[] { typeGroup });
+
+                GetDependencyObjectsRecursive(objList_specificType, allObjectsPerType, adapter);
+            }
+
+            return allObjectsPerType;
+        }
+
+        private static void GetDependencyObjectsRecursive<T>(this IEnumerable<T> objects, Dictionary<Type, List<IBHoMObject>> gatheredDependecies, IBHoMAdapter adapter ) where T : IBHoMObject
```
Comment (FraserGreenroyd) [inline Adapter_Engine/Query/GetDependencyRecursive.cs:66]:
```suggestion
        private static void GetDependencyObjectsRecursive<T>(this IEnumerable<T> objects, Dictionary<Type, List<IBHoMObject>> gatheredDependecies, IBHoMAdapter adapter) where T : IBHoMObject
```

File: Adapter_Engine/Query/GetDependencySortedObjects.cs [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+namespace BH.Engine.Adapter
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Push Support methods                      ****/
+        /***************************************************/
+        // These are support methods required by other methods in the Push process.
+
+        [Description("Recursively gets all the dependencies of the input objects. Then, the dependency objects and input objects are grouped by type." +
```
Comment (FraserGreenroyd) [inline Adapter_Engine/Query/GetDependencySortedObjects.cs:42]:
```suggestion
        [Description("Recursively gets all the dependencies of the input objects. Then, the dependency objects and input objects are grouped by type. " +
```

OR

```suggestion
        [Description("Recursively gets all the dependencies of the input objects. Then, the dependency objects and input objects are grouped by type.\n" +
```

File: Adapter_Engine/Query/GetDependencySortedObjects.cs [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+              "Finally, the groups are sorted by dependency order. " +
+              "The resulting collection can be used as an input to invoke CRUD methods in the correct dependency order.")]
+        [Input("objects", "Objects to group and sort by dependency order. The dependency of these objects will also be gathered recursively and included in the output.")]
+        [Input("bHoMAdapter", "The DependencyTypes that define the order of the output will be gathered from this Adapter instance.")]
```
Comment (FraserGreenroyd) [inline Adapter_Engine/Query/GetDependencySortedObjects.cs:46]: I think this description could be improved to explain what is needing to be input here, in addition to how things will be defined from it.

File: Adapter_oM/Adapter_oM.csproj [hunk 9]
```diff
@@ -16,7 +16,7 @@
   </PropertyGroup>
 
   <Target Name="PostBuild" AfterTargets="PostBuildEvent">
-    <Exec Command="xcopy &quot;$(TargetDir)$(TargetFileName)&quot;  &quot;C:\ProgramData\BHoM\Assemblies&quot; /Y" />
+    <Exec Command="xcopy &quot;$(TargetDir)$(TargetFileName)&quot;  &quot;C:\ProgramData\BHoM\Assemblies&quot; /Y&#xD;&#xA;&#xD;&#xA;Exit 0" />
```
Comment (FraserGreenroyd) [inline Adapter_oM/Adapter_oM.csproj:19]: 👀 Exit?

File: BHoM_Adapter/BHoM_Adapter.csproj [hunk 10]
```diff
@@ -69,6 +69,9 @@
       <Private>false</Private>
       <SpecificVersion>false</SpecificVersion>
     </Reference>
+    <Reference Include="Reflection_Engine">
+      <HintPath>..\..\..\..\..\..\ProgramData\BHoM\Assemblies\Reflection_Engine.dll</HintPath>
```
Comment (FraserGreenroyd) [inline BHoM_Adapter/BHoM_Adapter.csproj:73]: 👀 At least it's not a random exit this time 😄

Comment (FraserGreenroyd) [inline BHoM_Adapter/BHoM_Adapter.csproj:73]:
```suggestion
      <HintPath>C:\ProgramData\BHoM\Assemblies\Reflection_Engine.dll</HintPath>
```

You'll also need `SpecificVersion` and `CopyLocal` to be false - best to let the bot pick that up

File: BHoM_Adapter/BHoM_Adapter.csproj [hunk 11]
```diff
@@ -16,7 +16,7 @@
   </PropertyGroup>
 
   <Target Name="PostBuild" AfterTargets="PostBuildEvent">
-    <Exec Command="xcopy &quot;$(TargetDir)$(TargetFileName)&quot;  &quot;C:\ProgramData\BHoM\Assemblies&quot; /Y" />
+    <Exec Command="xcopy &quot;$(TargetDir)$(TargetFileName)&quot;  &quot;C:\ProgramData\BHoM\Assemblies&quot; /Y&#xD;&#xA;&#xD;&#xA;Exit 0" />
```
Comment (FraserGreenroyd) [inline BHoM_Adapter/BHoM_Adapter.csproj:19]: 👀

File: BHoM_Adapter_Tests/BHoM_Adapter_Tests.csproj [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+  </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\Adapter_Engine\Adapter_Engine.csproj" />
+    <ProjectReference Include="..\Adapter_oM\Adapter_oM.csproj" />
+    <ProjectReference Include="..\BHoM_Adapter\BHoM_Adapter.csproj" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Reference Include="Analytical_oM">
```
Comment (FraserGreenroyd) [inline BHoM_Adapter_Tests/BHoM_Adapter_Tests.csproj:26]: Run all of these through the bot

File: BHoM_Adapter_Tests/BHoM_Adapter_Tests.csproj [hunk 13]
```diff
@@ -0,0 +1,58 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
```
Comment (FraserGreenroyd) [inline BHoM_Adapter_Tests/BHoM_Adapter_Tests.csproj:4]: Should probably target `.NetStandard2.0` in line with the adapter itself?

File: BHoM_Adapter_Tests/PushTests.cs [hunk 14]
```diff
@@ -0,0 +1,84 @@
+using BH.Adapter.Tests;
```
Comment (FraserGreenroyd) [inline BHoM_Adapter_Tests/PushTests.cs:1]: Missing copyright header

File: BHoM_Adapter_Tests/Create/RandomObjects.cs [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+                objs.Add(obj);
+            }
+
+            return objs;
+        }
+    }
+
+
+    public class TestFragment : IFragment
```
Comment (FraserGreenroyd) [inline BHoM_Adapter_Tests/Create/RandomObjects.cs:88]: 👀 this should be its own file

File: BHoM_Adapter_Tests/StructuralAdapter.cs [hunk 16]
```diff
@@ -0,0 +1,85 @@
+﻿
```
Comment (FraserGreenroyd) [inline BHoM_Adapter_Tests/StructuralAdapter.cs:1]: Copyright header

File: Structure_AdapterModules/Structure_AdapterModules.csproj [hunk 17]
```diff
@@ -17,7 +17,7 @@
   </PropertyGroup>
 
   <Target Name="PostBuild" AfterTargets="PostBuildEvent">
-    <Exec Command="xcopy &quot;$(TargetDir)$(TargetFileName)&quot;  &quot;C:\ProgramData\BHoM\Assemblies&quot; /Y" />
+    <Exec Command="xcopy &quot;$(TargetDir)$(TargetFileName)&quot;  &quot;C:\ProgramData\BHoM\Assemblies&quot; /Y&#xD;&#xA;&#xD;&#xA;Exit 0" />
```
Comment (FraserGreenroyd) [inline Structure_AdapterModules/Structure_AdapterModules.csproj:20]: 👀 exit exit exit

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
See comments in the files changed for specific parts.

Regarding the testing - we already have a very nice framework set up using the `.ci/code` folder that we use for the tests such as versioning, serialisation, and null-handling. From what I can see in that the tests are doing within `BHoM_Adapter_Tests` here, all of that could be placed within the same framework quite happily, with a `Verify` method which can then be incorporated into the bot, which would allow us to then perform the tests on sub-toolkit adapters in addition to protecting the framework of the core adapter :smile:

File: BHoM_Adapter/BHoM_Adapter.csproj (hunk 11 above)
Comment (alelom) [reply BHoM_Adapter/BHoM_Adapter.csproj]: https://github.com/BHoM/BHoM_Adapter/pull/332/commits/42a17732ebd1c136b61919a23f5f575e5d5ff774

File: Structure_AdapterModules/Structure_AdapterModules.csproj (hunk 17 above)
Comment (alelom) [reply Structure_AdapterModules/Structure_AdapterModules.csproj]: https://github.com/BHoM/BHoM_Adapter/pull/332/commits/42a17732ebd1c136b61919a23f5f575e5d5ff774

File: Adapter_Engine/Query/BaseTypes.cs (hunk 3 above)
Comment (alelom) [reply Adapter_Engine/Query/BaseTypes.cs:54]: A failing TryGetType sets the `out` parameter to null, so the list needs to be initialized for the rest of the code.

File: Adapter_Engine/Query/BaseTypes.cs (hunk 1 above)
Comment (alelom) [reply Adapter_Engine/Query/BaseTypes.cs:45]: Agreed, placed temporarily here during development. I will move it and raise PR.

File: BHoM_Adapter_Tests/PushTests.cs (hunk 14 above)
Comment (alelom) [reply BHoM_Adapter_Tests/PushTests.cs:24]: ca5a982c8e7b6edd799dfc666c171dfee6addda6

File: BHoM_Adapter_Tests/Create/RandomObjects.cs (hunk 15 above)
Comment (alelom) [reply BHoM_Adapter_Tests/Create/RandomObjects.cs]: 87fb96088bc9051e4cd1fed9bbe30728a6a5e519

File: Adapter_Engine/Query/GetDependencySortedObjects.cs (hunk 8 above)
Comment (alelom) [reply Adapter_Engine/Query/GetDependencySortedObjects.cs:45]: Updated description.

**Review by IsakNaslundBh — COMMENTED** [review]
small comment

**Comment by IsakNaslundBh** [comment]
Generally happy with the functionality now. From my testing on a couple of adapters it is working great.

Also, generally happy with the code, see minor comments above regarding the latest commits.

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance

**Comment by alelom** [comment]
@BHoMBot check required

File: Adapter_Engine/Adapter_Engine.csproj [hunk 18]
```diff
@@ -2,8 +2,8 @@
 
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
-    <AssemblyVersion>6.0.0.0</AssemblyVersion>
-    <Description>https://github.com/BHoM/BHoM_Adapter</Description>
+    <AssemblyVersion>6.1.0.0</AssemblyVersion>
```
Comment (FraserGreenroyd) [inline Adapter_Engine/Adapter_Engine.csproj:5]:
```suggestion
    <AssemblyVersion>6.0.0.0</AssemblyVersion>
```

File: Adapter_oM/Adapter_oM.csproj [hunk 19]
```diff
@@ -2,8 +2,8 @@
 
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
-    <AssemblyVersion>6.0.0.0</AssemblyVersion>
-    <Description>https://github.com/BHoM/BHoM_Adapter</Description>
+    <AssemblyVersion>6.1.0.0</AssemblyVersion>
```
Comment (FraserGreenroyd) [inline Adapter_oM/Adapter_oM.csproj:5]:
```suggestion
    <AssemblyVersion>6.0.0.0</AssemblyVersion>
```

File: BHoM_Adapter/BHoM_Adapter.csproj [hunk 20]
```diff
@@ -2,8 +2,8 @@
 
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
-    <AssemblyVersion>6.0.0.0</AssemblyVersion>
-    <Description>https://github.com/BHoM/BHoM_Adapter</Description>
+    <AssemblyVersion>6.1.0.0</AssemblyVersion>
```
Comment (FraserGreenroyd) [inline BHoM_Adapter/BHoM_Adapter.csproj:5]:
```suggestion
    <AssemblyVersion>6.0.0.0</AssemblyVersion>
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

**Comment by alelom** [comment]
@BHoMBot check required

**Comment by alelom** [comment]
@BHoMBot check project-compliance

**Comment by alelom** [comment]
@BHoMBot check project-compliance

**Comment by alelom** [comment]
@BHoMBot check required

**Comment by FraserGreenroyd** [comment]
@IsakNaslundBh @alelom - Discussed with you both today but commenting for transparency and authorisation for @IsakNaslundBh that any compliance failures related SOLELY to the `.ci` folder are ok to be dispensated if I'm not around to provide it 😄

**Review by IsakNaslundBh — DISMISSED** [review]
Re-tested and happy with this functionality! Great to get this in!

**Comment by IsakNaslundBh** [comment]
@BHoMBot check copyright-compliance

File: .ci/code/BHoM_Adapter_Tests/PushTests.cs [hunk 21]
```diff
@@ -0,0 +1,180 @@
+/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2022, the respective contributors. All rights reserved.
```
Comment (IsakNaslundBh) [inline .ci/code/BHoM_Adapter_Tests/PushTests.cs:3]:
```suggestion
 * Copyright (c) 2015 - 2023, the respective contributors. All rights reserved.
```

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance
@BHoMBot check required

**Review by IsakNaslundBh — APPROVED** [review]
re-approving after updating copyright-year

**Comment by FraserGreenroyd** [comment]
@BHoMBot check ready-to-merge

**Comment by IsakNaslundBh** [comment]
@BHoMBot this is a DevOps instruction. I am authorising dispensation to be granted on check ref. 10483867458

**Comment by IsakNaslundBh** [comment]
Autorising the ProjectCOmpliance on the basis that only the .ci folder is affected, as agreed in comment from @FraserGreenroyd above

**Comment by IsakNaslundBh** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**